source/backend/api/Areas/DocumentGeneration/DocumentGenerationController.cs

```diff
@@ -54,17 +55,50 @@ public async Task<IActionResult> GetSupportedDocumentTypes()
         }
 
         /// <summary>
-        /// Uploads the passed document.
+        /// Uploads the passed document as a template.
         /// </summary>
         [HttpPost("upload_template")]
-        [HasPermission(Permissions.DocumentAdd)]
+        //[HasPermission(Permissions.GenerateDocuments)]
         [ProducesResponseType(typeof(ExternalResult<string>), 200)]
         [SwaggerOperation(Tags = new[] { "document-generation" })]
         public async Task<IActionResult> UploadTemplate([FromForm] IFormFile file)
         {
             var result = await _documentGenerationService.UploadFileTemplate(file);
             return new JsonResult(result);
         }
+
+        /// <summary>
+        /// Renders the given template with the request data and returns the result as wrapped encoded base64 file.
+        /// </summary>
+        [HttpPost("render_template/download-wrapped")]
+        //[HasPermission(Permissions.GenerateDocuments)]
+        [ProducesResponseType(typeof(ExternalResult<FileDownload>), 200)]
+        [SwaggerOperation(Tags = new[] { "document-generation" })]
+        public async Task<IActionResult> UploadTemplateAndDownloadWrapepd([FromBody] DocumentGenerationRequest request)
+        {
+            var result = await _documentGenerationService.GenerateDocument(request.TemplateType, request.TemplateData);
+            return new JsonResult(result);
+        }
+
+        /// <summary>
+        /// Renders the given template with the request data and returns the result as a file.
+        /// </summary>
+        [HttpPost("render_template/download")]
+        //[HasPermission(Permissions.GenerateDocuments)]
+        [ProducesResponseType(typeof(FileContentResult), 200)]
+        [SwaggerOperation(Tags = new[] { "document-generation" })]
+        public async Task<IActionResult> UploadTemplateAndDownload([FromBody] DocumentGenerationRequest request)
+        {
+            var result = await _documentGenerationService.GenerateDocument(request.TemplateType, request.TemplateData);
+            if (result?.Payload == null)
+            {
+                return new NotFoundResult();
+            }
+
+            byte[] fileBytes = System.Convert.FromBase64String(result.Payload.FilePayload);
+            return new FileContentResult(fileBytes, result.Payload.Mimetype) { FileDownloadName = result.Payload.FileName };
```

Comment:
See my other comment around file streams, this will have a large impact on the memory usage of the backend pods.

Reply:
There are two API's now. One does return a direct download link for the stream while the other one converts it to a base64 string with more information for the encoding.

Reply:
Might be worth thinking through the usage of that. One api will be usable across the application while the other will only be useable in areas where we can guarantee the file size is small. I think small file size guarantees may be possible for document template uploads/downloads, is that the use case that you were expecting for that API?

Reply:
Ideally, we would have the front end download directly from the backend API using the stream and setting the correct headers. That way we would be leveraging the browser's capabilities to stream data as well as the backend's. However, we are doing extra processing on the front end that might be preventing that (I am not 100% sure it does, but if not we should switch to the streaming API).

From the looks on the front end is because we cannot guarantee that the file is available, so we do some checks before prompting the download. We could have this revisited once we have that flag on the backend to know if the file has been processed and create a download link on the fly once the file has been fetched.